### PR TITLE
Add documentation on dynamism in StableHLO

### PR DESCRIPTION
**Pull request by GleasonK**
Added `dynamism.md` with information on dynamic shape handling in StableHLO.

A few other cleanups found while adding documentation:
- Added examples to argument refinement and canonicalize dynamism.
- Ordered passes alphabetically
- Rename `StablehloCreateCompatibilityExpanderPatterns` -> `StablehloCompatibilityExpanderPatterns`

### CODE REVIEW

File: docs/dynamism.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+[refine-arguments]:https://openxla.org/stablehlo/generated/stablehlo_passes#-stablehlo-refine-arguments
+[refine-shapes]:https://openxla.org/stablehlo/generated/stablehlo_passes#-stablehlo-refine-shapes
+
+## Example: How is dynamism useful, and how can I use it?
+
+Dynamism has lots of uses, here we'll mainly focus in on the common use case for
+Shape Polymorphism - creating a flexible exported model representation,
+generally used to represent dynamic batch size or sequence length.
+
+### Dynamic add_one model
```
Comment (sdasgup3) [inline docs/dynamism.md:124]: Same header name as the following one. Did you mean to say `Static add_one model`?

Comment (GleasonK) [reply docs/dynamism.md]: good catch, fixed. "Static add_one model"

File: stablehlo/transforms/StablehloCompatibilityExpander.cpp
Comment (ghpvnist) [inline stablehlo/transforms/StablehloCompatibilityExpander.cpp:1]: Let's use llvm [style guide](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements) here.

**Review by ghpvnist — APPROVED** [review]
Mostly nits, LGTM otherwise.

**Pull request merged**